spec/src/main/asciidoc/bulkhead.asciidoc

```diff
@@ -71,8 +71,20 @@ public Future<Connection> serviceA() {
 
 When using the thread pool approach, when a request cannot be added to the waiting queue, `BulkheadException` will be thrown.
 
+==== Lifecycle
+
+Bulkhead needs to needs to maintain some state between invocations: the number of currently running executions, or the queue of waiting executions.
+This state is a singleton, irrespective of the lifecycle of the bean that uses the `@Bulkhead` annotation.
+In other words, the state is associated with the bean class, not with its instances.
+For example, if the bean is `@RequestScoped` and therefore there is a new instance of the bean for each request, all instances share the same bulkhead state.
+
+More specifically: each method of each class that uses a bulkhead has its own state.
+If a subclass inherits a `@Bulkhead` method, then the subclass has its own, distinct, bulkhead state.
```

Comment:
I found this part surprising.

If a method is inherited by a subclass, why should it receive a distinct bulkhead state for the subclass? I has assumed that each method declaration would have a separate state, rather than each method declaration and subclass combination having a separate state.

Do you see advantages to defining it one way or the other?

I would see having one state per method declaration as consistent with the way we use annotations to apply fault tolerance policies to methods. When deciding which policies to apply, you must find the declaration of the method which is to be run and then consider:
 * annotations applied to the method
 * annotations applied to the class which declares the method
 * annotations applied to superclasses of the class which declares the method (because our annotations have `@Inherited`)

You ignore:
 * annotations applied to any method that the method overrides
 * annotations applied to any subclass of the class which declares the method

Reply:
I think it was you who pointed out that subclasses should have their own circuit breakers (applies to bulkhead as well): https://github.com/eclipse/microprofile-fault-tolerance/issues/479#issuecomment-561197781

But you raise very good points. I'll try to address them (tomorrow).

Reply:
Heh, when I was writing this, it did ring a few bells.

Unfortunately, I suspect whichever way round we do it, we might catch someone out :(

I think on balance, I'm leaning towards the state being per method declaration for the reasons set out above. It's also fairly easy to override the method and delegate to `super` if you do want a separate bulkhead state.

Reply:
I didn't forget about this, I just had a ton of other work. I got back to it today, and changed to your suggestion. I agree that tying the state to the method declaration makes most sense.

I'm wondering if something happens to a `@CircuitBreaker` method if a subclass which _doesn't_ override it adds a `@CircuitBreaker` annotation. I'd say no, but I'd also say that this should be specified somewhere for all annotations. Same thing with configuration overrides -- can I override configuration of a `@CircuitBreaker` method on a subclass, even if that subclass doesn't override the method? This is something that should be specified for all annotations. (I'm not sure if it is or not, didn't check :-) )